src/dataflow/src/logging/materialized.rs

```diff
@@ -263,6 +392,80 @@ pub fn construct<A: Allocate>(
+                                                    lower_bound: 0,
+                                                    length: bounds.len(),
+                                                }];
+                                                row_packer
+                                                    .push(Datum::from(metric.meta.name.as_str()));
+                                                row_packer.push(Datum::from(chrono_timestamp));
+                                                row_packer.push_dict(labels);
+                                                row_packer.push(Datum::from(sum));
+                                                row_packer.push(Datum::from(total_count));
+                                                let bounds = bounds.into_iter().map(Datum::from);
+                                                row_packer.push_array(dims, bounds).expect(
+                                                    "Mismatch in histogram array dimensions",
+                                                );
+                                                let counts = counts.into_iter().map(Datum::Int64);
+                                                row_packer.push_array(dims, counts).expect(
+                                                    "Mismatch in histogram array dimensions",
+                                                );
+                                                (
+                                                    row_packer.finish_and_reuse(),
+                                                    &mut metrics_histos_session,
+                                                )
+                                            }
+                                        };
+                                        session.give((row.clone(), time_ms, 1));
+                                        session.give((row, time_ms + retain_for, -1));
+                                    }
+                                    // Expire the metadata of a metric reading when the reading
+                                    // would expire, but refresh its lifetime when we get another
+                                    // reading referencing that metadata:
+                                    let meta = metric.meta;
+                                    let meta_expiry =
+                                        active_metrics.get(&meta).copied().unwrap_or(0);
+                                    if meta_expiry <= time_ms {
+                                        let row = meta.as_packed_row();
+                                        metrics_meta_session.give((row.clone(), time_ms, 1));
+                                        metrics_meta_session.give((row, time_ms + retain_for, -1));
+                                        active_metrics
+                                            .insert(meta, time_ms + retain_for as Timestamp);
+                                    }
```

Comment:
This seems likely to create blips where metadata disappears and reappears? Do we have any other alternatives for this?

Reply:
Agreed, this is not as useful for metadata as for the regular metrics. I'll see if we can revoke the entries manually once they need to go away.